### ISSUE TITLE
[9.x.x] Nestjs Cache Module Circular Dependency Error

### ISSUE THREAD

**Issue by tugascript** [nestjs/nest#9895]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

On adding the cache module to the main app imports will throw a circular dependency error:

```
[Nest] 23092  - 07/08/2022, 7:31:10 PM   ERROR [ExceptionHandler] A circular dependency has been detected inside CacheModule. Please, make sure that each side of a bidirectional relationships are decorated with "forwardRef()".
Error: A circular dependency has been detected inside CacheModule. Please, make sure that each side of a bidirectional relationships are decorated with "forwardRef()".
    at NestContainer.addProvider (/home/afonsobarracha/Documents/projects/project-example/node_modules/@nestjs/core/injector/container.js:108:19)
    at DependenciesScanner.insertProvider (/home/afonsobarracha/Documents/projects/project-example/node_modules/@nestjs/core/scanner.js:199:35)
    at /home/afonsobarracha/Documents/projects/project-example/node_modules/@nestjs/core/scanner.js:103:18
    at Array.forEach (<anonymous>)
    at DependenciesScanner.reflectProviders (/home/afonsobarracha/Documents/projects/project-example/node_modules/@nestjs/core/scanner.js:102:19)
    at DependenciesScanner.scanModulesForDependencies (/home/afonsobarracha/Documents/projects/project-example/node_modules/@nestjs/core/scanner.js:83:18)
    at DependenciesScanner.scan (/home/afonsobarracha/Documents/projects/project-example/node_modules/@nestjs/core/scanner.js:28:9)
    at /home/afonsobarracha/Documents/projects/project-example/node_modules/@nestjs/core/nest-factory.js:95:17
    at Function.asyncRun (/home/afonsobarracha/Documents/projects/project-example/node_modules/@nestjs/core/errors/exceptions-zone.js:22:13)
    at NestFactoryStatic.initialize (/home/afonsobarracha/Documents/projects/project-example/node_modules/@nestjs/core/nest-factory.js:94:13)
``` 

### Minimum reproduction code

https://github.com/tugascript/nestjs-cache-error

### Steps to reproduce

1. "yarn install"
2. "yarn start:dev"

### Expected behavior

To load the cache without a circular dependency error.

### Package

- [ ] I don't know. Or some 3rd-party package
- [X] <code>@nestjs/common</code>
- [ ] <code>@nestjs/core</code>
- [ ] <code>@nestjs/microservices</code>
- [ ] <code>@nestjs/platform-express</code>
- [ ] <code>@nestjs/platform-fastify</code>
- [ ] <code>@nestjs/platform-socket.io</code>
- [ ] <code>@nestjs/platform-ws</code>
- [ ] <code>@nestjs/testing</code>
- [ ] <code>@nestjs/websockets</code>
- [ ] Other (see below)

### NestJS version

9.0.1

### Packages versions

```json
{
  "name": "project-example",
  "version": "0.0.1",
  "description": "",
  "author": "",
  "private": true,
  "license": "UNLICENSED",
  "scripts": {
    "prebuild": "rimraf dist",
    "build": "nest build",
    "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
    "start": "nest start",
    "start:dev": "nest start --watch",
    "start:debug": "nest start --debug --watch",
    "start:prod": "node dist/main",
    "lint": "eslint \"{src,apps,libs,test}/**/*.ts\" --fix",
    "test": "jest",
    "test:watch": "jest --watch",
    "test:cov": "jest --coverage",
    "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
    "test:e2e": "jest --config ./test/jest-e2e.json"
  },
  "dependencies": {
    "@nestjs/common": "^9.0.1",
    "@nestjs/core": "^9.0.1",
    "@nestjs/platform-express": "^9.0.1",
    "cache-manager": "^4.0.1",
    "reflect-metadata": "^0.1.13",
    "rimraf": "^3.0.2",
    "rxjs": "^7.5.5"
  },
  "devDependencies": {
    "@nestjs/cli": "^9.0.0",
    "@nestjs/schematics": "^9.0.1",
    "@nestjs/testing": "^9.0.1",
    "@types/cache-manager": "^4.0.1",
    "@types/express": "^4.17.13",
    "@types/jest": "28.1.4",
    "@types/node": "^16.11.43",
    "@types/supertest": "^2.0.12",
    "@typescript-eslint/eslint-plugin": "^5.30.5",
    "@typescript-eslint/parser": "^5.30.5",
    "eslint": "^8.19.0",
    "eslint-config-prettier": "^8.5.0",
    "eslint-plugin-prettier": "^4.2.1",
    "jest": "28.1.2",
    "prettier": "^2.7.1",
    "source-map-support": "^0.5.21",
    "supertest": "^6.2.4",
    "ts-jest": "28.0.5",
    "ts-loader": "^9.3.1",
    "ts-node": "^10.8.2",
    "tsconfig-paths": "4.0.0",
    "typescript": "^4.7.4"
  },
  "jest": {
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "rootDir": "src",
    "testRegex": ".*\\.spec\\.ts$",
    "transform": {
      "^.+\\.(t|j)s$": "ts-jest"
    },
    "collectCoverageFrom": [
      "**/*.(t|j)s"
    ],
    "coverageDirectory": "../coverage",
    "testEnvironment": "node"
  },
  "packageManager": "yarn@3.2.1"
}

```

### In which operating systems have you tested?

- [ ] macOS
- [ ] Windows
- [X] Linux

### Other

OS is Fedora 36

**Comment by c58** [comment]
I have the same problem

**Comment by chunghha** [comment]
I have same problem too. And, no luck by replacing cache-manager-redis-store with cache-manager-ioredis.

<img width="461" alt="Screen Shot 2022-07-08 at 5 02 41 PM" src="https://user-images.githubusercontent.com/24458075/178075909-96ae6504-1d45-4f60-abd3-f6c2c01abbc8.png">

**Comment by micalevisk** [comment]
looks like a issue on `registerAsync` from the configurable module builder feature because  
 `CacheModule.register({ useFactory })` works while `CacheModule.registerAsync({ useFactory })` doesn't

**Comment by NeutronScott12** [comment]
Yep, the same problem has happened with me, as well, and as @micalevisk though I am having the same response with register as well.

**Comment by Innei** [comment]
I run into this problem now, there is a way to use `registerAsync`?

**Comment by micalevisk** [comment]
guys, saying that you have found this issue doesn't help to solve it. We are aware of this bug.

I didn't manage to investigate it better (I'm not using this feature so it isn't a priority to me)

---

Here's the source of `registerAsync` of `CacheModule`:

https://github.com/nestjs/nest/blob/34cb6001c30b3e2d2a9570bc87c6bd78a8c49524/packages/common/cache/cache.module.ts#L47-L56

and the source of `super.registerAsync`:

https://github.com/nestjs/nest/blob/34cb6001c30b3e2d2a9570bc87c6bd78a8c49524/packages/common/module-utils/configurable-module.builder.ts#L229-L248

PRs are more than welcomed :smile_cat:

**Comment by tugascript** [comment]
> guys, saying that you have found this issue as well doesn't help to solve it. We are aware of this bug.
> 
> I didn't manage to investigate it better (I'm not using this feature so it isn't a priority to me)
> 
> Here's the source of `registerAsync` on `CacheModule`:
> 
> https://github.com/nestjs/nest/blob/34cb6001c30b3e2d2a9570bc87c6bd78a8c49524/packages/common/cache/cache.module.ts#L47-L56
> 
> and the source of `super.registerAsync`:
> 
> https://github.com/nestjs/nest/blob/34cb6001c30b3e2d2a9570bc87c6bd78a8c49524/packages/common/module-utils/configurable-module.builder.ts#L229-L248
> 
> PRs are more than welcomed smile_cat

Took me way to much time to notice that there was a missing if statement in extraProviders :') Added a new PR with a ternary operator where the if statement was missing.

**Comment by micalevisk** [comment]
yea I just I found that that circular error appears when the `providers` array has some `undefined` element

**Comment by micalevisk** [comment]
> I am having the same response with register as well

@NeutronScott12 please share your code. The PR didn't touch the `register` method

**Comment by suitables** [comment]
Hey, here is the tmp solution for someone who can't wait new release!

`CacheModule.registerAsync({
      isGlobal: true,
      extraProviders: [],
      useFactory: async (configService: ConfigService): Promise<any> => ({
        store: redisStore,
        ...configService.get<any>('service.redis', {}),
        ttl: configService.get('cache.ttl'),
      }),
      inject: [ConfigService],
    }),`

**Comment by kamilmysliwiec** [comment]
Fixed in 9.0.2
